### PR TITLE
IMP schema resolution with default values support

### PR DESCRIPTION
**Pull request by ngavrysh**

### CODE REVIEW

File: lang/c/src/avro/schema.h [hunk 1]
```diff
@@ -45,6 +45,18 @@ avro_schema_t avro_schema_record(const char *name, const char *space);
...
+/*
+ * Default values
+ */
+json_t* avro_schema_record_field_default(const avro_schema_t schema, const char* field_name);
+
+char* avro_get_default_string_value(json_t* json_default);
+int avro_get_default_int_value(json_t* json_default);
+float avro_get_default_float_value(json_t* json_default);
+double avro_get_default_double_value(json_t* json_default);
+int8_t avro_get_default_bool_value(json_t* json_default);
```
Comment (ngavrysh) [inline lang/c/src/avro/schema.h]: Only primitive values for now. Will extend this in next commits.

File: lang/c/src/value-read-resolution.c [hunk 2]
```diff
@@ -0,0 +1,628 @@
...
+    st_table* reader_fields_table = st_init_strtable();
+    
+    // loop writers fields
+    for(i = 0; i<writers_fields_count; ++i)
+    {
+        avro_value_t field;
+        check(rval, avro_value_get_by_index(source, i, &field, NULL));
+        char* field_name =  avro_schema_record_field_name(record_schema_writer, i);
+        
+        int reades_index = avro_schema_record_field_get_index(record_schema_reader, field_name);
```
Comment (drudim) [inline lang/c/src/value-read-resolution.c]: `reader_index`?

File: lang/c/src/value-read-resolution.c [hunk 3]
```diff
@@ -0,0 +1,628 @@
...
+    if ( (rtype == AVRO_UNION) && (wtype == AVRO_UNION))
+    {
+        return resolve_unions(reader, source, dest);
+    }
+    else if ((rtype == AVRO_UNION) && (wtype != AVRO_UNION))
+    {
+        // reader is union, writer is not
+        int64_t  branch_count_dest;
+        branch_count_dest = avro_schema_union_size(rschema);
+        int i = 0;
```
Comment (drudim) [inline lang/c/src/value-read-resolution.c]: do you need this?

File: lang/c/src/value-read-resolution.c [hunk 4]
```diff
@@ -0,0 +1,628 @@
...
+        }
+
+        check(rval, avro_value_set_branch(source, discriminant, &branch));
+        check(rval, avro_value_get_current_branch(source, &branch));
+        return read_value_with_resolution(reader, &branch, dest);
+    }
+    else
+    {
+        // Shouldn't be here
+        avro_set_error("Cannot resolve unions.");
```
Comment (drudim) [inline lang/c/src/value-read-resolution.c:485]: is it possible to have such case?

File: lang/c/src/value-read-resolution.c [hunk 5]
```diff
@@ -0,0 +1,628 @@
...
+        avro_set_error("Invalid union discriminant value: (%d)",
+                       discriminant);
+        return EINVAL;
+    }
+    
+    check_prefix(rval,
+                 avro_value_set_branch(source, discriminant, &branch_source),
+                 "Cannot set current branch");
+    avro_schema_t source_branch_schema = avro_value_get_schema(&branch_source);
+    int i = 0;
```
Comment (drudim) [inline lang/c/src/value-read-resolution.c]: do you need this?

**Review by drudim — COMMENTED** [review]
I am not the right person to review `C` code. Scott, can I ask you to check this?

File: lang/c/src/value-read-resolution.c (hunk 5 above)
Comment (ngavrysh) [reply lang/c/src/value-read-resolution.c]: yes, C99 doesn't allow `for(int i = 0 ...)`

File: lang/c/src/avro/io.h [hunk 6]
```diff
@@ -126,6 +131,8 @@ int avro_file_reader_close(avro_file_reader_t reader);
 
 int
 avro_file_reader_read_value(avro_file_reader_t reader, avro_value_t *dest);
+int
+avro_file_reader_read_value_with_resolution(avro_file_reader_t r, avro_value_t *source, avro_value_t *value);
```
Comment (srstrickland) [inline lang/c/src/avro/io.h]: Maybe to keep things parallel use `dest` not `value`?

File: lang/c/src/avro/schema.h (hunk 1 above)
Comment (srstrickland) [reply lang/c/src/avro/schema.h]: in C99 you can `#include <stdbool.h>` and use `bool` ...

File: lang/c/src/value-read-resolution.c [hunk 7]
```diff
@@ -0,0 +1,628 @@
...
+    if(reader_fields_table)
+    {
+        // cleanup
+        st_free_table(reader_fields_table);
+    }
+    return 0;
+}
+
+// Simple macro to check if dest field is union before setting the value.
+#define avro_resolve_source(reader, source, dest, type) \
```
Comment (srstrickland) [inline lang/c/src/value-read-resolution.c]: Macros are pure evil. But if you insist, I think it's good practice to use `ALL_CAPS` for their names, so they aren't masquerading to the casual observer as a function call, which they are not.

File: lang/c/src/value-read-resolution.c [hunk 8]
```diff
@@ -0,0 +1,628 @@
...
+        {
+            return avro_resolve_source(reader, source, dest, type_null);
+        }
+            
+        case AVRO_STRING:
+        {
+            return avro_resolve_source(reader, source, dest, type_string);
+        }
+            
+        case AVRO_ARRAY:
```
Comment (srstrickland) [inline lang/c/src/value-read-resolution.c]: Why no braces here?

File: lang/c/src/value-read-resolution.c [hunk 9]
```diff
@@ -0,0 +1,628 @@
...
+        {
+            return avro_resolve_source(reader, source, dest, type_enum);
+        }
+            
+        case AVRO_FIXED:
+        {
+            return avro_resolve_source(reader, source, dest, type_fixed);
+        }
+            
+        case AVRO_MAP:
```
Comment (srstrickland) [inline lang/c/src/value-read-resolution.c]: ditto braces... pick one (yes or no) and stick with it

File: lang/c/src/value-read-resolution.c (hunk 3 above)
Comment (srstrickland) [reply lang/c/src/value-read-resolution.c]: Don't need the initialization (`= 0;`) part, or don't need the re-init in the for loop.  Probably redundant to have it both places.

File: lang/c/src/value-read.c [hunk 10]
```diff
@@ -34,11 +35,11 @@
  * reset in avro_value_read itself).
  */
 
-static int
+int
```
Comment (srstrickland) [inline lang/c/src/value-read.c]: why turn these global?

**Review by srstrickland — COMMENTED** [review]
mostly minor stuff.  it might be worthwhile to pump this stuff through valgrind if you haven't already to verify that we don't have any memory leaks, since this all low-level C stuff and it's easy for something to slip through.

**Comment by ngavrysh** [comment]
@srstrickland I've checked this with profiler and nothing seems to be leaking.
Ready for review.

**Review by srstrickland — APPROVED** [review]

**Pull request merged**